zipline/pipeline/graph.py

```diff
@@ -70,7 +70,9 @@ def offset(self):
 
             Max number of extra rows needed by any term depending on `input`
             minus
-            Number of extra rows needed by `term`.
+            Number of extra rows needed by `input` for `term`
+            minus
+            Number of extra rows needed by `term`
```

Comment:
> Can you explain a bit more about why this is necessary? I don't have much intuition for why this is correct just from reading it.

Since we now have windowed inputs, such as Returns, we get this sort of two-layered lookback effect where the Correlation factors need `X` days of extra Returns, but Returns itself needs `Y` days of extra close prices. The result is that when calculating Return's offset, we need to adjust for the fact that we need `X` extra days of a term for which we previously thought we didn't need any.

For example, if we have a Correlation factor with `returns_length = 3` and `correlation_length = 5`, then we end up with (3 - 1) + (5 - 1) = 6 extra days (that is, 6 days before our start_date) of close price in order to calculate the (5 - 1) = 4 extra days of Returns. Previously, Return's offset in this case would be 6 (the max number of extra days needed for close price) minus 2 (the number of extra rows of close price needed by Returns) = 4. If Returns has a `date_offset` (given in `SimplePipelineEngine._mask_and_dates_for_term`) of 2, the total offset for Returns is 6, bringing us back to our start_date. However, we wanted 2 extra days of Returns, so we subtract from the offset the "Number of extra rows needed by term", which for Returns is 2.

Hopefully this example helps, otherwise I can try to elaborate in a more general sense.